Realm/RLMObjectBase.mm

```diff
@@ -58,26 +59,56 @@ - (instancetype)init {
     return self;
 }
 
+static id RLMValidatedObjectForProperty(id obj, RLMProperty *prop, RLMSchema *schema) {
+    if (!RLMIsObjectValidForProperty(obj, prop)) {
+        // check for object or array of properties
+        if (prop.type == RLMPropertyTypeObject) {
+            // for object create and try to initialize with obj
+            RLMObjectSchema *objSchema = schema[prop.objectClassName];
+            return [[objSchema.objectClass alloc] initWithValue:obj schema:schema];
+        }
+        else if (prop.type == RLMPropertyTypeArray && [obj conformsToProtocol:@protocol(NSFastEnumeration)]) {
+            // for arrays, create objects for each element and return new array
+            RLMObjectSchema *objSchema = schema[prop.objectClassName];
+            RLMArray *objects = [[RLMArray alloc] initWithObjectClassName: objSchema.className standalone:YES];
+            for (id el in obj) {
+                [objects addObject:[[objSchema.objectClass alloc] initWithValue:el schema:schema]];
+            }
+            return objects;
+        }
+
+        // if not convertible to prop throw
+        @throw RLMException([NSString stringWithFormat:@"Invalid value '%@' for property '%@'", obj, prop.name]);
+    }
+    return obj;
+}
+
 - (instancetype)initWithValue:(id)value schema:(RLMSchema *)schema {
     self = [self init];
+    NSArray *properties = _objectSchema.properties;
     if (NSArray *array = RLMDynamicCast<NSArray>(value)) {
-        // validate and populate
-        array = RLMValidatedArrayForObjectSchema(array, _objectSchema, schema);
-        NSArray *properties = _objectSchema.properties;
+        if (array.count != properties.count) {
+            @throw RLMException(@"Invalid array input. Number of array elements does not match number of properties.");
+        }
         for (NSUInteger i = 0; i < array.count; i++) {
-            [self setValue:array[i] forKeyPath:[properties[i] name]];
+            [self setValue:RLMValidatedObjectForProperty(array[i], properties[i], schema) forKeyPath:[properties[i] name]];
         }
     }
     else {
         // assume our object is an NSDictionary or a an object with kvc properties
-        NSDictionary *dict = RLMValidatedDictionaryForObjectSchema(value, _objectSchema, schema);
-        for (NSString *name in dict) {
-            id val = dict[name];
-            // strip out NSNull before passing values to standalone setters
-            if (val == NSNull.null) {
-                val = nil;
+        NSDictionary *defaultValues = nil;
```

Comment:
This code is quite similar to what's in `RLMValidateValueForObjectSchema`, could you write a function?